### PR TITLE
Windows: Handle CreateProcess() fail due to extra large stack.

### PR DESCRIPTION
**Pull request by cher-nov**
Now it's handled as `MemoryLimitExceeded` error.
This should close klenin/cats-judge#44.

### CODE REVIEW

File: libspawner/src/win32/runner.cpp [hunk 1]
```diff
@@ -353,10 +353,8 @@ thread_return_t runner::async_body(thread_param_t param) {
 runner::runner(const std::string &program, const options_class &options)
     : base_runner(program, options)
     , running_thread(INVALID_HANDLE_VALUE)
-    , init_semaphore(INVALID_HANDLE_VALUE) {
-
-    init_semaphore = CreateSemaphore(NULL, 0, 10, NULL);
-    ZeroMemory(&process_info, sizeof(process_info));
```
Comment (klenin) [inline libspawner/src/win32/runner.cpp]: Why did you delete this?

File: CMakeLists.txt [hunk 2]
```diff
@@ -80,13 +80,10 @@ set(SOURCES
     src/spawner_pcms2.cpp
 )
 
-set(LIBRARIES libspawner)
+set(LIBRARIES "libspawner")
```
Comment (klenin) [inline CMakeLists.txt]: It seems most of existing code does not use quotes.
I would prefer this way unless there is compelling reason to quote.

File: CMakeLists.txt [hunk 3]
```diff
@@ -80,13 +80,10 @@ set(SOURCES
...
 
 if(WIN32)
-    find_library(USERENV userenv)
-    if(USERENV)
-        set(LIBRARIES ${LIBRARIES} ${USERENV})
-    endif()
+    set(LIBRARIES ${LIBRARIES} "userenv")
```
Comment (klenin) [inline CMakeLists.txt]: Did you test MSVC?

File: libspawner/src/win32/runner.cpp [hunk 4]
```diff
@@ -136,6 +139,27 @@ void runner::restore_original_environment(const runner::env_vars_list_t& origina
     }
 }
 
+void runner::analyze_executable(const std::string& exe_path) {
+    LOADED_IMAGE exe_image;
+    std::unique_ptr<char> exe_path_rw(_strdup(exe_path.c_str()));
+    if (!MapAndLoad(exe_path_rw.get(), NULL, &exe_image, FALSE, TRUE)) {return;}
```
Comment (klenin) [inline libspawner/src/win32/runner.cpp]: Remove braces.

File: libspawner/src/win32/runner.cpp [hunk 5]
```diff
@@ -136,6 +139,27 @@ void runner::restore_original_environment(const runner::env_vars_list_t& origina
     }
 }
 
+void runner::analyze_executable(const std::string& exe_path) {
+    LOADED_IMAGE exe_image;
+    std::unique_ptr<char> exe_path_rw(_strdup(exe_path.c_str()));
```
Comment (klenin) [inline libspawner/src/win32/runner.cpp]: Why all this?

File: libspawner/src/win32/runner.cpp [hunk 6]
```diff
@@ -136,6 +139,27 @@ void runner::restore_original_environment(const runner::env_vars_list_t& origina
...
+    if (!MapAndLoad(exe_path_rw.get(), NULL, &exe_image, FALSE, TRUE)) {return;}
+
+    stack_exceeds_1gb = 0x40000000 < exe_image.FileHeader->OptionalHeader.SizeOfStackReserve;
```
Comment (klenin) [inline libspawner/src/win32/runner.cpp]: What if stack is below 1Gb but above memory limit?

File: libspawner/src/win32/runner.cpp [hunk 7]
```diff
@@ -480,6 +506,7 @@ void runner::run_process() {
         WaitForSingleObject(process_info.hProcess, INFINITE);//may stuck here
         return;
     }
+    analyze_executable(program);
```
Comment (klenin) [inline libspawner/src/win32/runner.cpp]: Either exit here if stack is too large or analyze only in case of error.

File: libspawner/src/win32/runner.cpp (hunk 1 above)
Comment (cher-nov) [inline libspawner/src/win32/runner.cpp]: I didn't delete anything, I just moved this into initializer list.
`ZeroMemory()` is worse than empty brace-initializer, I believe.

File: libspawner/src/win32/runner.cpp (hunk 5 above)
Comment (cher-nov) [inline libspawner/src/win32/runner.cpp]: WinAPI. It requests `LPSTR`, which is non-constant modifyable C-string (i.e. `char*`).
`c_str()` returns `const char*`. Passing this into function will cause UB (although MSVC is quietly silent).

File: CMakeLists.txt (hunk 2 above)
Comment (cher-nov) [inline CMakeLists.txt]: I've quoted it just because library names below are quoted too.

File: libspawner/src/win32/runner.cpp (hunk 7 above)
Comment (cher-nov) [inline libspawner/src/win32/runner.cpp]: > exit here if stack is too large
We never know when stack is large enough to cause `CreateProcess()` failure. We only could make an assumption that if it exceeds system guaranteed maximum, then an error may occur.
But I definitely should analyze only in case of error, sorry.

File: CMakeLists.txt (hunk 3 above)
Comment (cher-nov) [inline CMakeLists.txt]: Ehm. Of course.

File: libspawner/src/win32/runner.cpp (hunk 6 above)
Comment (cher-nov) [inline libspawner/src/win32/runner.cpp]: Ehm. I didn't understand the question.
It at least equal to the memory limit of the vast majority of problems, because default stack size on Windows is 1 MiB.
